### PR TITLE
A tree-based (non-overlapping) solver

### PR DESCRIPTION
**Pull request by EgorKraevTransferwise**
## Context

A tree-based solver for wise-pizza, producing non-overlapping, complete coverage of the whole dataset by segments
Doesn't support time series yet

## Checklist
- [x ] Change meets or does not compromise the [Baseline Security Requirements](https://transferwise.atlassian.net/wiki/spaces/EKB/pages/434929973/Baseline+Security+Requirements)

### CODE REVIEW

File: wise_pizza/transform.py [hunk 1]
```diff
@@ -71,19 +71,28 @@ def inverse_transform_weight(self, w: np.ndarray, x: np.ndarray) -> np.ndarray:
...
     def inverse_transform_mean(self, x: np.ndarray) -> np.ndarray:
-        return np.maximum(
-            0.0, np.exp(np.minimum(x, np.log(self.max_inverse))) - self.offset
-        )
+        if self.cap_inverse:
+            return np.maximum(
+                0.0, np.exp(np.minimum(x, np.log(self.max_inverse))) - self.offset
+            )
+        else:
+            np.maximum(0.0, np.exp(x) - self.offset)
```
Comment (AlxdrPolyakov) [inline wise_pizza/transform.py:95]: Probably missing "return"

Comment (AlxdrPolyakov) [inline wise_pizza/transform.py]: Probably missing "return"

File: wise_pizza/transform.py [hunk 2]
```diff
@@ -71,19 +71,28 @@ def inverse_transform_weight(self, w: np.ndarray, x: np.ndarray) -> np.ndarray:
...
-        self.max_inverse = max_inverse
+        self.cap_inverse = cap_inverse
+        if cap_inverse:
+            self.max_inverse = 0.0
+        else:
+            self.max_inverse = None
 
     def transform_mean(self, x: np.ndarray) -> np.ndarray:
+        if self.cap_inverse:
+            self.max_inverse = np.maximum(self.max_inverse, 2 * x.max())
```
Comment (AlxdrPolyakov) [inline wise_pizza/transform.py:86]: We transform max_inverse in that function, but then we don't use it there. Is it the best place for this?

File: wise_pizza/solve/tree.py [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+
+
+def build_tree(root: ModelNode, num_leaves: int, max_depth: Optional[int] = 1000):
+    # TODO: modify this to also accept max_depth
+    for _ in range(num_leaves - 1):
+        best_node = get_best_subtree_result(root, max_depth)
+        if best_node.error_improvement > 0:
+            best_node.children = best_node._best_submodels
+        else:
+            break
```
Comment (AlxdrPolyakov) [inline wise_pizza/solve/tree.py:173]: if we want to end in case of error_improvement <= 0, it may be more stable to do "return" I think

**Review by AlxdrPolyakov — APPROVED** [review]

File: wise_pizza/transform.py (hunk 2 above)
Comment (EgorKraevTransferwise) [reply wise_pizza/transform.py:86]: This is actually only used in the time series bit, can we come back to it in the next iteration on time series?

File: wise_pizza/solve/tree.py (hunk 3 above)
Comment (EgorKraevTransferwise) [reply wise_pizza/solve/tree.py:173]: Why? In this case I think they're exactly equivalent, no?

File: wise_pizza/transform.py (hunk 2 above)
Comment (AlxdrPolyakov) [reply wise_pizza/transform.py:86]: Sure

File: wise_pizza/transform.py (hunk 1 above)
Comment (EgorKraevTransferwise) [reply wise_pizza/transform.py]: Nice catch, thanks! :)

File: wise_pizza/solve/tree.py (hunk 3 above)
Comment (AlxdrPolyakov) [reply wise_pizza/solve/tree.py:173]: Yes, of course they are equivalent in this case, it's just that if there is nothing after the loop, I don't see the point of doing a break anyway. But it doesn't matter and we can leave it as it is.

**Pull request merged**